pkg/shared/families/infofinder/sshtopology/scanner.go

```diff
@@ -0,0 +1,364 @@
+	conf := c.(types.ScannersConfig) // nolint:forcetypeassert
+	return &Scanner{
+		name:       ScannerName,
+		logger:     logger.Dup().WithField("scanner", ScannerName),
+		config:     conf.SSHTopology,
+		resultChan: resultChan,
+	}
+}
+
+// nolint:cyclop,gocognit
+func (s *Scanner) Run(sourceType utils.SourceType, userInput string) error {
+	go func() {
+		s.logger.Debugf("Running with input=%v and source type=%v", userInput, sourceType)
+		retResults := types.ScannerResult{
+			ScannerName: ScannerName,
+		}
+
+		// Validate this is an input type supported by the scanner,
+		// otherwise return skipped.
+		if !s.isValidInputType(sourceType) {
+			s.sendResults(retResults, nil)
+			return
+		}
+
+		var errs []error
+		homeUserDirs, err := getHomeUserDirs(userInput)
+		if err != nil {
+			// Collect the error and continue.
+			errs = append(errs, fmt.Errorf("failed to get home user dirs: %v", err))
+		}
+		s.logger.Debugf("Found home user dirs %+v", homeUserDirs)
+
+		errorsChan := make(chan error)
+		fingerprintsChan := make(chan []types.Info)
+
+		go func() {
+			for fingerprints := range fingerprintsChan {
+				retResults.Infos = append(retResults.Infos, fingerprints...)
+			}
+		}()
+
+		go func() {
+			for e := range errorsChan {
+				errs = append(errs, e)
+			}
+		}()
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			if sshDaemonKeysFingerprints, err := s.getSSHDaemonKeysFingerprints(userInput); err != nil {
+				errorsChan <- fmt.Errorf("failed to get ssh daemon keys: %v", err)
+			} else {
+				fingerprintsChan <- sshDaemonKeysFingerprints
+			}
+		}()
+
+		for i := range homeUserDirs {
+			dir := homeUserDirs[i]
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				if sshPrivateKeysFingerprints, err := s.getSSHPrivateKeysFingerprints(dir); err != nil {
+					errorsChan <- fmt.Errorf("failed to get ssh private keys: %v", err)
+				} else {
+					fingerprintsChan <- sshPrivateKeysFingerprints
+				}
+			}()
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				if sshAuthorizedKeysFingerprints, err := s.getSSHAuthorizedKeysFingerprints(dir); err != nil {
+					errorsChan <- fmt.Errorf("failed to get ssh authorized keys: %v", err)
+				} else {
+					fingerprintsChan <- sshAuthorizedKeysFingerprints
+				}
+			}()
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				if sshKnownHostsFingerprints, err := s.getSSHKnownHostsFingerprints(dir); err != nil {
+					errorsChan <- fmt.Errorf("failed to get ssh known hosts: %v", err)
+				} else {
+					fingerprintsChan <- sshKnownHostsFingerprints
+				}
+			}()
+		}
+
+		wg.Wait()
+		close(errorsChan)
+		close(fingerprintsChan)
+
+		retErr := errors.Join(errs...)
+		if retErr != nil {
+			retResults.Error = retErr
+		}
+
+		if len(retResults.Infos) > 0 && retResults.Error != nil {
+			// Since we have findings we want to share what we've got and only prints the errors here.
```

Comment:
Since we have findings, we want to share what we've got and only **print** the errors here.